core/templates/services/auth.service.spec.ts

```diff
@@ -0,0 +1,130 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Unit tests for AuthService.
+ */
+
+import { TestBed } from '@angular/core/testing';
+import { AngularFireAuth } from '@angular/fire/auth';
+import { of } from 'rxjs';
+import { marbles } from 'rxjs-marbles';
+
+import { AuthService } from 'services/auth.service';
+import { MockAngularFireAuth } from 'tests/unit-test-utils';
+
+
+describe('Auth service', () => {
+  const setUpSystemUnderTest = (
+      idTokenSource$ = of(null)): [AngularFireAuth, AuthService] => {
+    const mockAngularFireAuth = new MockAngularFireAuth(idTokenSource$);
+
+    TestBed.configureTestingModule({
+      providers: [
+        AuthService,
+        {provide: AngularFireAuth, useValue: mockAngularFireAuth},
+      ]
+    });
+
+    return [TestBed.inject(AngularFireAuth), TestBed.inject(AuthService)];
+  };
+
+  it('should resolve when sign out succeeds', async() => {
+    const [angularFireAuth, authService] = setUpSystemUnderTest();
+
+    spyOn(angularFireAuth, 'signOut').and.resolveTo();
+
+    await expectAsync(authService.signOutAsync()).toBeResolvedTo();
+  });
+
+  it('should reject when sign out fails', async() => {
+    const [angularFireAuth, authService] = setUpSystemUnderTest();
+
+    spyOn(angularFireAuth, 'signOut').and.rejectWith(new Error('fail'));
+
+    await expectAsync(authService.signOutAsync()).toBeRejectedWithError('fail');
+  });
+
+  it('should unsubscribe from source after calling ngOnDestroy', marbles(m => {
+    const sourceIdTokens = m.hot('   ^---');
+    const authServiceSubscription = '^-! ';
+
+    const [, authService] = setUpSystemUnderTest(sourceIdTokens);
+    // Calls ngOnDestroy() after 2 frames have elapsed in the marble diagram.
+    // Frame 0 begins at the ^ character, so we check for an unsubscribe
+    // (represented by the ! character) 2 frames after it.
+    m.scheduler.schedule(() => authService.ngOnDestroy(), 2);
+
+    m.expect(sourceIdTokens).toHaveSubscriptions(authServiceSubscription);
```

Comment:
Hm, sorry, I don't think I quite understand this. In what way is this "checking for unsubscription from source"?

Reply:
Ah, subscription here means `.subscribe()` and `.unsubscribe()`. The diagram begins at the subscription and corresponds to frame 0. The `!` character represents the `.unsubscribe()`, which gets called by `ngOnDestroy()`.

Reply:
I've made updates to the description to help, let me know if things are still unclear.

Reply:
Sorry, still don't get it. I get that a destroy action is happening. But then why does sourceIdTokens still "have" authServiceSubscription after the subscription has ended?

(I.e. I would expect a test along the lines of "has subscriptions before t=2, doesn't have subscriptions after t = 2")

Reply:
`AuthService`'s subscription lasts for a "sub-section" of source's overall lifetime. `ngOnDestroy` is the action that ends the subscription, hence why the '!' appears in the subscription diagram.

I've added a few extra characters to help make this a bit more obvious.

Reply:
`AuthService`'s subscription lasts for a _subset_ of source's overall lifetime. `ngOnDestroy` is the action that ends the subscription, hence why `!` appears in the subscription diagram.

I've added a few extra characters to the diagram and one more sentence to highlight the lifetime expectations being made here.

Reply:
@srijanreddy98 for ideas on how to make this clearer?